### PR TITLE
ddl: fix the issue that the length of the index is 0.

### LINKED ISSUE #10644: The used storage engine can't index column ‘xxxx’
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

**I encountered this error while creating the table,**
```
CREATE TABLE `MCH_STORE_LABEL` (
  `LABEL_MAP_ID` int(11) NOT NULL,
  `STORE_ID` int(11) NOT NULL ,
  `LABEL_ID` int(11) NOT NULL ,
  `CREATE_BY` varchar(20) NOT NULL ,
  `CREATE_DATE` datetime NOT NULL,
  `UPDATE_BY` varchar(20) DEFAULT NULL,
  `LAST_MODIFIED` datetime NOT NULL ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`LABEL_MAP_ID`)
) ENGINE=Innodb DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ;

```
**Then perform this alteration:**
```
ALTER TABLE `MCH_STORE_LABEL`
ADD INDEX `INDEX_STORE_ID`(`STORE_ID`) USING BTREE;
```

**error log of tidb.log:**

```
2019/05/30 10:12:09.366 session.go:1638: [info] [CRUCIAL OPERATION] con:150591 schema_ver:868 cur_db                                                                                                             :ystpay ALTER TABLE `ystpay`.`MCH_STORE_LABEL`
ADD INDEX `INDEX_STORE_ID`(`STORE_ID`) USING BTREE (by root@192.168.0.58).
2019/05/30 10:12:09.366 tidb.go:157: [info] RollbackTxn for ddl/autocommit error.
2019/05/30 10:12:09.367 session.go:892: [warning] con:150591 schema_ver:868 session error:
[ddl:1167]The used storage engine can't index column 'STORE_ID'
github.com/pingcap/errors.AddStack
        /home/jenkins/workspace/build_tidb_master/go/pkg/mod/github.com/pingcap/errors@v0.11.0/error                                                                                                             s.go:174
github.com/pingcap/parser/terror.(*Error).GenWithStackByArgs
        /home/jenkins/workspace/build_tidb_master/go/pkg/mod/github.com/pingcap/parser@v0.0.0-201902                                                                                                             28070002-74e8cffabf28/terror/terror.go:233
github.com/pingcap/tidb/ddl.buildIndexColumns
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/ddl/index.go:63
github.com/pingcap/tidb/ddl.(*ddl).CreateIndex
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/ddl/ddl_api.go:2632
github.com/pingcap/tidb/ddl.(*ddl).AlterTable
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/ddl/ddl_api.go:1547
github.com/pingcap/tidb/executor.(*DDLExec).executeAlterTable
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/ddl.go:291
github.com/pingcap/tidb/executor.(*DDLExec).Next
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/ddl.go:104
github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/adapter.go                                                                                                             :298
github.com/pingcap/tidb/executor.(*ExecStmt).Exec
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/adapter.go                                                                                                             :244
github.com/pingcap/tidb/session.runStmt
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/session/tidb.go:206
github.com/pingcap/tidb/session.(*session).executeStatement
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/session/session.go:                                                                                                             889
github.com/pingcap/tidb/session.(*session).execute
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/session/session.go:                                                                                                             957
github.com/pingcap/tidb/session.(*session).Execute
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/session/session.go:                                                                                                             910
github.com/pingcap/tidb/server.(*TiDBContext).Execute
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/driver_tidb.                                                                                                             go:246
github.com/pingcap/tidb/server.(*clientConn).handleQuery
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:1017
github.com/pingcap/tidb/server.(*clientConn).dispatch
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:745
github.com/pingcap/tidb/server.(*clientConn).Run
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:582
github.com/pingcap/tidb/server.(*Server).onConn
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/server.go:40                                                                                                             5
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1333
{
  "currDBName": "ystpay",
  "id": 150591,
  "status": 2,
  "strictMode": true,
  "user": {
    "Username": "root",
    "Hostname": "192.168.0.58",
    "CurrentUser": false,
    "AuthUsername": "root",
    "AuthHostname": "%"
  }
}
2019/05/30 10:12:09.367 conn.go:600: [warning] con:150591 dispatch error:
id:150591, addr:192.168.0.58:29639 status:2, collation:utf8_general_ci, user:root
"ALTER TABLE `ystpay`.`MCH_STORE_LABEL` \r\nADD INDEX `INDEX_STORE_ID`(`STORE_ID`) USING BTREE"
[ddl:1167]The used storage engine can't index column 'STORE_ID'
github.com/pingcap/errors.AddStack
        /home/jenkins/workspace/build_tidb_master/go/pkg/mod/github.com/pingcap/errors@v0.11.0/error                                                                                                             s.go:174
github.com/pingcap/parser/terror.(*Error).GenWithStackByArgs
        /home/jenkins/workspace/build_tidb_master/go/pkg/mod/github.com/pingcap/parser@v0.0.0-201902                                                                                                             28070002-74e8cffabf28/terror/terror.go:233
github.com/pingcap/tidb/ddl.buildIndexColumns
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/ddl/index.go:63
github.com/pingcap/tidb/ddl.(*ddl).CreateIndex
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/ddl/ddl_api.go:2632
github.com/pingcap/tidb/ddl.(*ddl).AlterTable
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/ddl/ddl_api.go:1547
github.com/pingcap/tidb/executor.(*DDLExec).executeAlterTable
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/ddl.go:291
github.com/pingcap/tidb/executor.(*DDLExec).Next
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/ddl.go:104
github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/adapter.go                                                                                                             :298
github.com/pingcap/tidb/executor.(*ExecStmt).Exec
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/executor/adapter.go                                                                                                             :244
github.com/pingcap/tidb/session.runStmt
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/session/tidb.go:206
github.com/pingcap/tidb/session.(*session).executeStatement
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/session/session.go:                                                                                                             889
github.com/pingcap/tidb/session.(*session).execute
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/session/session.go:                                                                                                             957
github.com/pingcap/tidb/session.(*session).Execute
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/session/session.go:                                                                                                             910
github.com/pingcap/tidb/server.(*TiDBContext).Execute
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/driver_tidb.                                                                                                             go:246
github.com/pingcap/tidb/server.(*clientConn).handleQuery
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:1017
github.com/pingcap/tidb/server.(*clientConn).dispatch
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:745
github.com/pingcap/tidb/server.(*clientConn).Run
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:582
github.com/pingcap/tidb/server.(*Server).onConn
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/server.go:40                                                                                                             5
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1333
```

2. What did you expect to see?

Why can't I add the index?
But here's the weird thing,I can add the index when I rebuild this table.So this bug may not fire every time.

3. What did you see instead?

I see the notice on navicat ：
The used storage engine can't index column 'STORE_ID'

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
Release Version: v3.0.0-beta-153-g63b49c9e4
Git Commit Hash: 63b49c9e42ec51022e6e4e385d17b202dcf76c6d
Git Branch: master
UTC Build Time: 2019-03-05 06:08:25
GoVersion: go version go1.11.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
Fix #10644 

### What is changed and how it works?
Add the type check when the length of the index is 0.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -61,7 +61,7 @@ func buildIndexColumns(columns []*model.ColumnInfo, idxColNames []*ast.IndexColN
 			return nil, errKeyColumnDoesNotExits.GenWithStack("column does not exist: %s", ic.Column.Name)
 		}
 
-		if col.Flen == 0 {
+		if col.Flen == 0 && (types.IsTypeChar(col.FieldType.Tp) || types.IsTypeVarchar(col.FieldType.Tp)) {
```
Comment (jackysp) [inline ddl/index.go]: could we use `IsString`? I'm not sure the type blob should check the length or not.

Comment (zimulala) [reply ddl/index.go]: @jackysp  `IsTypeBlob` is different errors, so I only use these two.

Comment (jackysp) [reply ddl/index.go]: OK.

**Comment by zimulala** [comment]
PTAL @bb7133 @crazycs520

**Pull request merged**